src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.inline.hpp

```diff
@@ -462,219 +544,262 @@ ShenandoahScanRemembered<RememberedSet>::mark_range_as_empty(HeapWord *addr, siz
-      if (is_dirty) {
-        size_t prev_card_index = card_index;
-        if (has_object) {
-          // Scan all objects that start within this card region.
-          size_t start_offset = _scc->get_first_start(card_index);
-          HeapWord *p = _rs->addr_for_card_index(card_index);
-          HeapWord *card_start = p;
-          HeapWord *endp = p + CardTable::card_size_in_words();
-          assert(!r->is_humongous(), "Process humongous regions elsewhere");
-
-          if (endp > end_of_range) {
-            endp = end_of_range;
-            next_card_index = end_card_index;
+      // [dirty_l, dirty_r] is a "maximal" closed interval range of dirty card indices:
+      // it may not be maximal if we are using the write_table, because of concurrent
+      // mutations dirtying the card-table. It may also not be maximal if an upper bound
+      // was established by the scan of the previous chunk.
+      const size_t dirty_l = cur_index + 1;   // record left end of dirty range (inclusive)
+      // Check that we identified a boundary on our left
+      assert(ctbm[dirty_l] == CardTable::dirty_card_val(), "First card in range should be dirty");
+      assert(dirty_l == start_card_index || use_write_table
+             || ctbm[dirty_l - 1] == CardTable::clean_card_val(),
+             "Interval isn't maximal on the left");
+      assert(dirty_r >= dirty_l, "Error");
+      assert(ctbm[dirty_r] == CardTable::dirty_card_val(), "Last card in range should be dirty");
+      // Record alternations, dirty run length, and dirty card count
+      NOT_PRODUCT(stats.record_dirty_run(dirty_r - dirty_l + 1);)
+
+      // Find first object that starts this range:
+      // [left, right) is a maximal right-open interval of dirty cards
+      HeapWord* left = _rs->addr_for_card_index(dirty_l);        // inclusive
+      HeapWord* right = _rs->addr_for_card_index(dirty_r + 1);   // exclusive
+      // Clip right to end_addr established above (still exclusive)
+      right = MIN2(right, end_addr);
+      assert(right <= region->top() && end_addr <= region->top(), "Busted bounds");
+      const MemRegion mr(left, right);
+
+      // TODO: cache so as to not call block_start() repeatedly
+      // on a very large object. Check: may be upper_bound below
+      // and its use above already takes care of this.
+      HeapWord* p = _scc->block_start(dirty_l);
+      oop obj = cast_to_oop(p);;
+
+      // PREFIX: The object that straddles into this range of dirty cards
+      // from the left may be subject to special treatment unless
+      // it is an object array.
+#define OBJ_MARK_IMPRECISE false   // indicates that objects are _always_ imprecisely marked:
+                                   // this is not true today, as GC closures are precise, but
+                                   // compiler & interpreter are imprecise.
+      if (p < left && !obj->is_objArray()) {
+        if (OBJ_MARK_IMPRECISE) {
```

Comment:
The code would be simpler if the case for handling `OBJ_MARK_IMPRECISE` were left out - perhaps moved to a descriptive comment? If we leave it in, could we name it `OBJ_MARK_ALWAYS_IMPRECISE` . The barriers do use imprecise card marks, so it's confusing to see this defined as `false`.

Reply:
Yes, I agree that the naming is confusing, plus the unnecessary code clutter.

I removed that arm and associated preprocessor def'n, and fixed up the comments.